### DISCUSSION TITLE
[SOLVED] Things to look out for in the registration process

### DISCUSSION THREAD

**Discussion by sammyskills** [codeigniter4/shield#343] [Ideas]
Hi guys,

I've been thinking about this for a while now, and I think it's best I share with you all. 

During registration _(when an email action is enabled)_, CI-Shield doesn't follow the commonly used process of sending a verification link which can be clicked from the user's email. While I think this is great, I think there are a few flaws we need to look at:

1. After a successful registration, users are redirected to the `auth/a/show` page. What happens when the activation email is not received by the user immediately, maybe due to poor network, etc, can the user retrieve the token page after the page is exited and session/cookie is deleted?

Here's a use-case scenario:
* User A registers but doesn't receive the activation email until after 4 hours due to poor network or maybe got distracted by something else
* User A turns off their device or simply closes the current browser which deletes the current session/cookie that makes the token verification page to show

At this point, when User A decides to come back to the website after retrieving the activation code from their email _(maybe with another browser)_, they are presented with the login page or registration page and the token verification page is no longer accessible. 
How can the user verify/activate their account?

2. I think there should be a way to resend the activation code, just in case a user doesn't get it at first trial.

**Comment by MGatner** [comment]
I think resending an activation code is an important feature, especially because identities are unique in the database so once a registration is started a user may not start over if the activation fails.

**Reply by sammyskills** [reply]
Yes, it is important. I can create a PR for that.

What about cases where the cookie/session for the `auth/a/show` is deleted or expires before the user gets the activation code in their email? 
As it stands, there is no way to get that page again, which in my opinion, is not good. We need to look at every possible way to make the registration process as simple as possible for users.

**Reply by MGatner** [reply]
The current process has been crafted with attention to security. I'm all in favor of making things easier for the end user, but not if it risks the integrity of the process. I think for now the focus should be on a meaningful way to restart the process, rather than adjusting the process itself.

**Reply by sammyskills** [reply]
I understand that we shouldn't compromise security for ease-of-use, but what do you suggest in such cases?

If a user doesn't get the activation code until after the session/cookie is deleted, how can the user continue the registration process?

**Reply by MGatner** [reply]
The likelihood of session expiration seems pretty low, so I don't think re-crafting the process to account for that is worth it. 

I think if a user goes to register and there is an existing, unconfirmed identity token, then it should be replaced so the user can start the process over again.

**Reply by sammyskills** [reply]
> The likelihood of session expiration seems pretty low, so I don't think re-crafting the process to account for that is worth it.

Well, this is not totally correct. I ran shield in a test environment and used different mobile devices to test. There are devices that kill all browser sessions immediately a user switches from the browser app to another app. So when a user comes back to the browser, it shows a 404 error page.
Another thing I noticed was that for devices that didn't kill the session, whenever a user returns back to the browser, it automatically reloads the page which causes a new activation code to be sent. For cases of poor network, this code will be sent multiple times, and the user might not be able get past the verification page.

> I think if a user goes to register and there is an existing, unconfirmed identity token, then it should be replaced so the user can start the process over again.

Then this means the user will have to go over the registration process again. I don't think users will find it funny IMO.

**Reply by MGatner** [reply]
Since you started this thread @kenjis has discovered some other issues with the activation process, so maybe a larger refactor is in order.

**Reply by sammyskills** [reply]
So, here are the current problems:

- **Problem 1**: _How to get the verification page after the session expires._
- **Problem 2**: _How to resend the activation code_
- **Problem 3**: _How to prevent unwanted sending of multiple activation codes_

And these are the solutions I'm proposing:

1. After successful registration, in the `auth/a/show` page, we can have a button `Resend code` that can be used to resend the code. _(Not sure if this is within the scope of shield, but it will be a nice addition to make the resending of codes time limited. That is, after it has been clicked, the user cannot click on the button again until after 60 seconds)_. This should fix **Problem 2.**
2. In cases where the session expires before the user got the code, there can be an information in the email that goes out stating `"If you can no longer access the verification page, please login with your registered details to generate a new activation code."`. 
So, when a user visits the login or register page with details of an unverified detail, the activation code is resent to the user and the user is redirected to the `auth/a/show` page to complete the verification process. This should fix **Problem 1**.
Also, I think we need to check for `session` in the login and registration pages, to verify if a user exists in a pending login state. If this is true, then the user can be redirected to the verification page to request for a new code. This should fix #349 
3. To prevent unwanted sending out of multiple activation emails, we can create a cookie/session when the first code is sent, and check for it's existence on page load. Whenever the code is resent by clicking on the `Resend code` button, we can also update this cookie/session. This should fix **Problem 3**.

Please, what do you think?

**Reply by lonnieezell** [reply]
I think a `Resend Email` feature is a good addition. I don't know that I would worry about a time limit. I think each email sent should overwrite the previous identity that was generated for the previous one so that only the most recent email would work. 

Second, I think we should refactor to not rely on the session anymore. Or at least not be the only solution. At these times we would have the action information stored in user identities so we can check for the existence of them and know what needs to happen. If an email verification action exists (and the system still wants verification) ensure that step is shown first. 

Thoughts?

**Reply by sammyskills** [reply]
I think this is the way to go!

**Reply by kenjis** [reply]
1. The session data is cache. If not available, we must check the database.
2. ~~The user state *inactive* is missing in the Session Authenticator now.~~
  See https://github.com/codeigniter4/shield/blob/bc85d79599ec7150f9bcad45d3622179919afe74/src/Authentication/Authenticators/Session.php#L40-L43

**Reply by kenjis** [reply]
> After successful registration, in the auth/a/show page, we can have a button Resend code that can be used to resend the code. (Not sure if this is within the scope of shield, but it will be a nice addition to make the resending of codes time limited. That is, after it has been clicked, the user cannot click on the button again until after 60 seconds). This should fix Problem 2.

The page is:
> Email Activation
We just sent an email to you with a code to confirm your email address. Copy that code and paste it below.

Are you suggesting that we add a resend code button to this page?
If a user has that page open, they should be able to see the email and send the code.
I don't understand the need for the button.

**Reply by sammyskills** [reply]
Yes, I think there should be a resend code button. 

As explained, a user may not get the code initially, so having that button should help in such cases.

**Reply by kenjis** [reply]
Oh, I see.

**Reply by pixobit** [reply]
The time limit is important, because if the email doesn't get delivered right away, the user might resend it again right away, this is common behavior, now if you get the first email, and the 2nd one is delayed, you will see it is not working anymore, so you would resend it again, but now you get the 2nd email, and so on. 
I'm saying this, because I did experience this before on some site, and it was frustrating. A time limit would avoid this scenario.

**Reply by kenjis** [reply]
Do you mean all codes should be valid if they are not expired?
It would certainly be more usable.

**Reply by pixobit** [reply]
You're right, that would be even more convenient

**Reply by warcooft** [reply]
> The time limit is important, because if the email doesn't get delivered right away, the user might resend it again right away, this is common behavior, now if you get the first email, and the 2nd one is delayed, you will see it is not working anymore, so you would resend it again, but now you get the 2nd email, and so on. 
> I'm saying this, because I did experience this before on some site, and it was frustrating. A time limit would avoid this scenario.

This scenario often occurs in the field and yes it is a nightmare. is there a plan to implement a time limit feature to avoid this scenario?

**Reply by datamweb** [reply]
Due to the flexibility of the shield, this possibility can be implemented by developers in a custom way.
Currently, there is no plan for implementation by the team.

**Comment by kenjis** [comment]
> The user state *inactive* is missing in the Session Authenticator now.

The current implementation see a *inactive* user as *pending*.
Both are restricted states. I looked into the code, there does not seem to be a need to strictly distinguish between the two.

I take back this opinon.

**Comment by pixobit** [comment]
What about multi device usage? Right now if I want to confirm the email address from a different device/browser, it would throw invalid code. The resend functionality doesn't really solve this issue, unless it's specifically mentioned for the user, otherwise it might cause confusion...

**Comment by pjsde** [comment]
> Second, I think we should refactor to not rely on the session anymore. Or at least not be the only solution. At these times we would have the action information stored in user identities so we can check for the existence of them and know what needs to happen. If an email verification action exists (and the system still wants verification) ensure that step is shown first.

I agree with @lonnieezell to not rely on the session anymore

**Reply by kenjis** [reply]
I sent a PR: https://github.com/codeigniter4/shield/pull/391
